open_spiel/python/pytorch/ppo.py

```diff
@@ -0,0 +1,344 @@
+
+    logits = self.actor(x)
+    probs = CategoricalMasked(logits=logits, masks=legal_actions_mask, mask_value=self.mask_value)
+    if action is None:
+      action = probs.sample()
+    return action, probs.log_prob(action), probs.entropy(), self.critic(x), probs.probs
+
+
+class PPOAtariAgent(nn.Module):
+  def __init__(self, num_actions, observation_shape, device):
+    super(PPOAtariAgent, self).__init__()
+    # Note: this network is intended for atari games, taken from https://github.com/vwxyzjn/ppo-implementation-details/blob/main/ppo_atari.py
+    self.network = nn.Sequential(
+      layer_init(nn.Conv2d(4, 32, 8, stride=4)),
+      nn.ReLU(),
+      layer_init(nn.Conv2d(32, 64, 4, stride=2)),
+      nn.ReLU(),
+      layer_init(nn.Conv2d(64, 64, 3, stride=1)),
+      nn.ReLU(),
+      nn.Flatten(),
+      layer_init(nn.Linear(64 * 7 * 7, 512)),
+      nn.ReLU(),
+    )
+    self.actor = layer_init(nn.Linear(512, num_actions), std=0.01)
+    self.critic = layer_init(nn.Linear(512, 1), std=1)
+    self.num_actions = num_actions
+    self.device = device
+    self.register_buffer("mask_value", torch.tensor(INVALID_ACTION_PENALTY))
+
+  def get_value(self, x):
+    return self.critic(self.network(x / 255.0))
+
+  def get_action_and_value(self, x, legal_actions_mask=None, action=None):
+    if legal_actions_mask is None:
+      legal_actions_mask = torch.ones((len(x), self.num_actions)).bool()
+    
+    hidden = self.network(x / 255.0)
+    logits = self.actor(hidden)
+    probs = CategoricalMasked(logits=logits, masks=legal_actions_mask, mask_value=self.mask_value)
+      
+    if action is None:
+      action = probs.sample()
+    return action, probs.log_prob(action), probs.entropy(), self.critic(hidden), probs.probs
+
+def legal_actions_to_mask(legal_actions_list, num_actions):
+  '''Convert a list of legal actions to a mask of size num actions with a 1 in a legal position'''
+  legal_actions_mask = torch.zeros((len(legal_actions_list), num_actions), dtype=torch.bool)
+  for i, legal_actions in enumerate(legal_actions_list):
+    legal_actions_mask[i, legal_actions] = 1
+  return legal_actions_mask
+
+class PPO(nn.Module):
+  def __init__(
+    self, 
+    input_shape, 
+    num_actions, 
+    num_players,
+    player_id=0,
+    num_envs=1,
+    steps_per_batch=128,
+    num_minibatches=4,
+    update_epochs=4,
+    learning_rate=2.5e-4, 
+    num_annealing_updates=None, 
+    gae=True,
+    gamma=0.99,
+    gae_lambda=0.95,
+    normalize_advantages=True,
+    clip_coef=0.2,
+    clip_vloss=True,
+    entropy_coef=0.01,
+    value_coef=0.5,
+    max_grad_norm=0.5,
+    target_kl=None,
+    device='cpu', 
+    writer=None, # Tensorboard SummaryWriter
+    agent_fn=PPOAtariAgent,
+    ):
+    super().__init__()
+
+    self.input_shape = input_shape
+    self.num_actions = num_actions
+    self.num_players = num_players
+    self.player_id = player_id
+    self.device = device
+
+    # Training settings
+    self.num_envs = num_envs
```

Comment:
Am I correct in assuming this is the number of consecutive environment steps in each batch element  ? 
I would suggest a more explicit name - num_env_steps_in_batch_elems. Or anything more elegant. ;)

Reply:
No - this is the number of environments that are run concurrently in each step. It's a tweak to PPO that we haven't seen elsewhere in OpenSpiel, but tends to have a significant impact on performance. We'll add a note in the PPO docstring.